addons/l10n_sa_edi/models/account_edi_format.py

```diff
@@ -0,0 +1,810 @@
+
+    def _l10n_sa_check_partner_missing_info(self, partner_id, is_commercial=True):
+        """
+            Helper function to check if ZATCA mandated partner fields are missing for a specified partner record
+        :param recordset partner_id: Partner record to check
+        :return: Missing partner fields
+        :rtype: list
+        """
+        missing = []
+        fields_to_check = [
+            ('l10n_sa_edi_building_number', _('Building Number')),
+            ('l10n_sa_edi_plot_identification', _('Plot Identification (4 digits)'), lambda v: len(str(v)) == 4),
+            ('l10n_sa_edi_neighborhood', _('Neighborhood')),
+            ('l10n_sa_additional_identification_scheme',
+             _('Additional Identification Scheme is required for commercial partners')),
+            ('l10n_sa_additional_identification_number',
+             _('Additional Identification Number is required for commercial partners')),
+            ('state_id', _('State / Country subdivision'))
+        ]
+        for field in fields_to_check:
+            field_value = partner_id[field[0]]
+            if not field_value or (len(field) == 3 and not field[2](field_value)):
+                missing.append(field[1])
+        return missing
+
+    def _l10n_sa_edi_is_required_for_invoice(self, invoice):
+        """
+            Determine whether a ZATCA EDI document needs to be generated for a given invoice
+        """
+        return invoice.is_sale_document() and invoice.country_code == 'SA'
+
+    def _l10n_sa_get_invoice_name(self, invoice):
+        """
+            Generate the name of the invoice XML file according to ZATCA business rules:
+            Seller Vat Number (BT-31), Date (BT-2), Time (KSA-25), Invoice Number (BT-1)
+        """
+        vat = invoice.company_id.partner_id.commercial_partner_id.vat
+        invoice_number = re.sub("[^a-zA-Z0-9 -]", "-", invoice.name)
+        invoice_date = self._l10n_sa_get_zatca_datetime(invoice.l10n_sa_confirmation_datetime)
+        return '%s_%s_%s.xml' % (vat, invoice_date.strftime('%Y%m%dT%H%M%S'), invoice_number)
+
+    def _l10n_sa_post_zatca_edi(self, invoice):  # no batch ensure that there is only one invoice
+        """
+            Post invoice to ZATCA and return a dict of invoices and their success/attachment
+        """
+
+        # Chain integrity check: chain head must have been REALLY posted, and did not time out
+        chain_head = invoice.journal_id._l10n_sa_get_last_posted_invoice()
+        if chain_head and chain_head != invoice and not chain_head.l10n_sa_chain_index:
+            return {
+                'error': f"ZATCA: Cannot post invoice while chain head ({chain_head.name}) has not been posted",
+                'blocking_level': 'error'
+            }
+
+        if not invoice.l10n_sa_confirmation_datetime:
+            invoice.l10n_sa_confirmation_datetime = fields.Datetime.now()
```

Comment:
@jco-odoo The field is used to generate the QR code, as well as the Unsigned XML content (before sending). Both of these happen once the invoice is posted, and the value of the field is reset every time the invoice is reset to draft then confirmed, since if an invoice was rejected, it needs to be fixed then submitted again

Reply:
Lines removes